spec/datadog/profiling/stack_recorder_spec.rb

```diff
@@ -545,6 +545,42 @@ def sample_allocation(obj)
             expect(heap_sample.values[:'heap-live-samples']).to eq(sample_rate * heap_sample_every)
           end
         end
+
+        it 'act on implicit frees from rb_gc_force_recycle' do
+          skip 'Behavior does not apply to current Ruby version' if RUBY_VERSION >= '3.1'
+
+          test_num_recycled_object = 60000
+
+          # We'll start by allocating strings that we'll sample for heap profiling then force recycle them.
+          test_num_recycled_object.times do |i|
+            object_to_be_recycled = "this is recycled string number #{i}"
+            sample_allocation(object_to_be_recycled)
+            described_class::Testing._native_gc_force_recycle(object_to_be_recycled)
+          end
+
+          allocation_line = __LINE__ - 4
+
+          test_num_non_recycled_object = 60000
+          live_objects = Array.new(test_num_non_recycled_object)
+
+          # We'll then allocate a whole lot of other strings and hope that at least some of them re-use
+          # the force-recycled slots rather than constantly asking for new ones.
+          test_num_non_recycled_object.times do |i|
+            live_objects[i] = "this is non-recycled string number #{i}"
+          end
+
+          # If we act on implicit frees, then we assume that even though some of the sampled objects
+          # ended up being re-used by non-recycled strings, we were able to detect this re-use and
+          # discard those tracked heap samples, ending with no matching recycled samples at all.
+          heap_samples_in_test_matcher = lambda { |sample|
+            (sample.values[:'heap-live-samples'] || 0) > 0 && sample.locations.any? do |location|
+              location.lineno == allocation_line && location.path == __FILE__
+            end
+          }
+
+          relevant_sample = heap_samples.find(&heap_samples_in_test_matcher)
+          expect(relevant_sample).to be nil
+        end
```

Comment:
I really like that we're adding coverage for this, even if it's awkward.

Having said that, I somewhat worry that in its current form this test is somewhat brittle, since it relies on allocating a lot and then hoping at least some of those 60k objects hit the issue.

I have some suggestions on how to change this, but I realize it may be a bit of work so... yeah, I guess up to you for how deep you want to go on this rabbit hole lol.

It seems to me that there's a few different cases that may be interesting to test:

1. That if we try to sample a recycled object, we will set its RUBY_FL_SEEN_OBJ_ID bit
2. That we'll correctly handle sampling a recycled object that was born on the spot of a tracked object (== that has the same id)
3. That we'll correctly clean up object tracking during serialization if we spot that a recycled object

For 1, one way that occurs to me of creating a recycled object is to allocate objects in a loop, get their object_id, and testing their `RUBY_FL_SEEN_OBJ_ID`. Eventually that should give us a recycled object.

For 2. and 3. I think we can modify the above to also track every object that gets created. Once we spot a recycled object, we could then either try to sample it to test 2. or try to flush the profile to test 3.

One trick for testing some of these weirder cases that worked well in the `CpuAndWallTimeWorker` and `ThreadContext` collectors was to introduce counters ("stats") to be able to observe/assert that we took the expected codepath for some weird situation.

Reply:
Went with a slightly different approach but I got the intended level of granularity and seems to work locally